docs/_content/ontwikkelaars/algemeen/authenticatie-autorisatie.md

```diff
@@ -74,10 +137,29 @@ Dit mechanisme zet zich door tot aan de zaaktypecatalogus.
 
 ![sequence](./_assets/authenticatie-autorisatie.png?raw=true)
 
+## Wijzigingen ten opzichte van de vorige vorm
+
+In een eerdere sprint werden de scope/zaaktypes claims opgenomen in het JWT
+wat verstuurd wordt van de (taak)applicatie naar de API. Dit is niet langer
```

Comment:
Het weghalen hiervan voorkomt wel dat het onmogelijk wordt om op gebruikersniveau scopes toe te wijzen. Iedereen heeft nu eigenlijk toegang tot de scopes van de applicatie.

De applicatie moet dus zelf bepalen wat een gebruiker wel of niet mag zoals je hieronder ook aangeeft.

In NLx gaan ze uit van autorisatie in de outway, waar de gebruiker bekend is, om scopes op te stellen. Ik denk dat een AC als dit vrij eenvoudig kan worden omgevormd om ook een gebruiker mee te nemen in het verhaal en specifiekere scopes terug te geven. **Echter betekent dat weer dat de componenten WEL nog Authorization headers aan moeten kunnen met claims!**

Reply:
Bij Utrecht is het zo dat ze geen SSO/ADFS hebben, maar gebruikers nog per taakapplicatie aangemaakt worden en permissies ingesteld. Je hebt echter wel een punt dat er misschien nog een centrale entiteit is die weet wat welke gebruiker mag. Ik weet niet of dan de taakapplicatie dit moet ophalen uit een AC of een ADFS o.i.d. om daar het eigen autorisatiemodel mee aan te drijven, of dat dit op API-niveau dichtgetimmerd moet worden.

In principe hebben we wel altijd gesteld dat we kijken naar de consumer, en niet naar de eindgebruiker die ervoor zorgt dat de consumer calls maakt.

De applicatie was ook al verantwoordelijk voor het mappen van rollen naar scopes, en een luie applicatiebouwer, die neemt gewoon altijd alle scopes op want die maakt enkel de API calls die een gebruiker volgens de UI/applicatiepermissies mag (aanname).

Ik zie niet echt een probleem dus.

Reply:
> 
> 
> Bij Utrecht is het zo dat ze geen SSO/ADFS hebben, maar gebruikers nog per taakapplicatie aangemaakt worden en permissies ingesteld. Je hebt echter wel een punt dat er misschien nog een centrale entiteit is die weet wat welke gebruiker mag. Ik weet niet of dan de taakapplicatie dit moet ophalen uit een AC of een ADFS o.i.d. om daar het eigen autorisatiemodel mee aan te drijven, of dat dit op API-niveau dichtgetimmerd moet worden.
> 
> In principe hebben we wel altijd gesteld dat we kijken naar de consumer, en niet naar de eindgebruiker die ervoor zorgt dat de consumer calls maakt.
> 
> De applicatie was ook al verantwoordelijk voor het mappen van rollen naar scopes, en een luie applicatiebouwer, die neemt gewoon altijd alle scopes op want die maakt enkel de API calls die een gebruiker volgens de UI/applicatiepermissies mag (aanname).
> 
> Ik zie niet echt een probleem dus.
Een taakapplicatie blijft altijd verantwoordelijk om een gebruiker (die al dan niet via SSO op een applicatie inlogt) te koppelen aan een bepaald intern autorisatieprofiel (waarin staat welke rechten bij dat profiel horen). Oplossingen als ADFS maar ook bijvoorbeeld een IAM oplossing met SCIM kunnen hoogstens een gebruiker aan een 'applicatie autorisatie profiel' koppelen, de taakapplicatie is verder verantwoordelijk dat alleen de acties die bij dat profiel horen, uitgevoerd kunnen worden.

Reply:
> Ik zie niet echt een probleem dus.

Maar je gaat claim-logica verwijderen uit het JWT. Het is dus niet meer mogelijk om applicaties gebruiker-specifieke claims mee te geven?!

Reply:
verduidelijkt via de getrapte-autorisatiepiramide :-)

Reply:
Joeri heeft hier wel een goed punt: de flexibiliteit om afhankelijk van de gebruiker en zijn rechten andere scopes mee te geven is weg hiermee. De client_id bepaalt volledig wat wel en niet mag. Dat lijkt me ongewenst. De zaaktype-specifieke scopes waren juist bedoeld om gebruikers op basis van specifieke rechten toegang te kunnen geven op zaken.

Reply:
in de vorige situatie was het theoretisch mogelijk dat een taakapplicatie voor een gebruiker bevraagt wat die mag (welke scopes) en dat werd in het JWT opgenomen. In de praktijk zie/zag ik dat eigenlijk niet gebeuren, want:

1. aanname centrale voorziening -> die moet gebouwd worden, bestaat niet, en lastiger verhaal in een context waarbij geen SSO bestaat
2. zonder centrale voorziening: betekent nog steeds dat het in de applicatie moet gemapped worden. Developers zijn lui, ik verwacht hier eigenlijk dat leveranciers hun eigen model behouden en gewoon altijd alle scopes gebruiken, want, lekker makkelijk en weg van minste weerstand.
3. in essentie is het nog steeds de client vertrouwen, omdat de server die checks niet kan dubbelchecken. dit lag toen ook al bij de taakapplicatie om inderdaad die vertaling te doen/te vragen
4. oude en nieuwe situatie autoriseren _applicaties_ op een ZRC/DRC/ZTC etc.
5. nieuwe situatie werkt preventief in plaats van oude situatie waar je op logging/audit moet steunen

ik zie op zich wel mogelijkheden om het AC uit te breiden waarbij een applicatie een claim meestuurt van 'dit is deze gebruiker (openid oid)', wat dan met gedetailleerde autorisaties in het AC kan gebouwd worden. dat wordt echter heel complex, en heel veel beheersoverlast
de realiteit blijkt te zijn dat autorisaties voor een individuele gebruiker ingesteld worden in de taakapplicatie in het eigen autorisatiemodel. je hebt echt een decoupling tussen dat niveau, en communicatie van de applicatie naar de APIs

Reply:
Ik sluit mij aan bij het verhaal van Sergei. Tijdens API-lab hebben we de driehoek met gelaagdheid besproken en het is het meest logische plaatje bij autorisaties:

![driehoek](https://raw.githubusercontent.com/VNG-Realisatie/gemma-zaken/2a6299207e7c527090ec2ef1840210401e6e6248/docs/_content/ontwikkelaars/algemeen/_assets/autorisaties.png)

Dat er voorheen individuele claims/scopes mogelijk waren per gebruiker betekent in de praktijk niet dat het ook zo gebruikt wordt. Het is logischer dat dit gebeurt binnen de taakapplicatie op een manier die los staat van claims en scopes.

In mijn [eerdere comment](https://github.com/VNG-Realisatie/gemma-zaken/pull/890/files#r274967782) was ik nog niet overtuigd, nu wel :)